### PR TITLE
Fix sticky nav conflicting with app banners

### LINKED ISSUE #10881: Sticky ad broken when in combination with iOS banner
1) The ad hides behind the iOS banner, leaving a gap

img_0433

2) When you scroll down, the ad has fixed whitespace of about 300px beneath it

img_0434

### PR DESCRIPTION
**Pull request by OliverJAsh**
I think we need to improve the quality of our experience for first time users, as this is crucial to their impression of our site.

There is a bug whereby the sticky nav conflicts with the app install banner. This results in a poor user experience for all our first time users who are using an iOS or Android device (tablet only, because the sticky nav is disabled on mobile). Not only that, we continue to show the app banner until after the fourth impression, plus many users who browse in incognito mode will get this experience for every page view. :cry: 

I recommend viewing the diff in no whitespace mode: https://github.com/guardian/frontend/pull/10914/files?w=1

Fixes #10881

### CODE REVIEW

File: static/src/javascripts/projects/common/utils/$.js [hunk 1]
```diff
@@ -8,6 +8,13 @@ define([
     _
 ) {
 
+    // Warning: side effect. This patches the bonzo module for use everywhere
+    bonzo.aug({
+        height: function () {
+            return this.dim().height;
+        }
+    });
+
```
Comment (OliverJAsh) [inline static/src/javascripts/projects/common/utils/$.js]: Not sure about this. I would prefer to extend the constructor/“class”, but bonzo unfortunately doesn't expose the constructor—instead they have this `bonzo.aug`.

File: static/src/javascripts/projects/common/modules/navigation/sticky.js [hunk 2]
```diff
@@ -302,32 +306,60 @@ define([
...
+                    this.$els.header.css({
+                        position: '',
+                        width: '',
+                        'margin-top': '',
+                        '-webkit-transform': '',
+                        '-ms-transform': '',
+                        'transform': '',
+                        'z-index': ''
+                    });
+
```
Comment (OliverJAsh) [inline static/src/javascripts/projects/common/modules/navigation/sticky.js]: This resets the above CSS declarations

File: static/src/javascripts/projects/common/modules/navigation/sticky.js [hunk 3]
```diff
@@ -228,9 +228,13 @@ define([
 
         this.setScrollDirection(scrollY);
 
+        var $appBanner = $('.site-message--ios, .site-message--android');
```
Comment (Calanthe) [inline static/src/javascripts/projects/common/modules/navigation/sticky.js]: I thought that we have a rule to use only one var?

File: static/src/javascripts/projects/common/modules/navigation/sticky.js [hunk 4]
```diff
@@ -228,9 +228,13 @@ define([
...
+        var appBannerHeight = $appBanner.height();
```
Comment (Calanthe) [inline static/src/javascripts/projects/common/modules/navigation/sticky.js]: It's nice that you use your own bonzo's height method but we have lots of `dim().height` in this file. I am afraid that now you will have to refactor it and use the `.height()` method.

**Comment by Calanthe** [comment]
Other than that it looks fine :+1: 

I am more than sure that I implemented 'app banner' fix some time ago, but don't know what happened to that code. Maybe it was removed after we turned off the sticky nav on mobile. 

The `sticky.js` is getting bigger and bigger. Did you find it easy to work with? Do you have any ideas how to refactor it so it may be less complicated?

**Comment by OliverJAsh** [comment]
> The sticky.js is getting bigger and bigger. Did you find it easy to work with? Do you have any ideas how to refactor it so it may be less complicated?

Yes, it's huge! There are many different code paths depending on content type, breakpoint, scroll position, if the navigation is locked or not, if the page is sensitive or not, and more. As an example, the code that simply just sets the header to fixed positioning appears duplicated four times! I think the complexity of the code probably suggests the design is too complicated, and because of that, we have too many edge cases to worry about.

I think we need to massively simplify the design.

File: static/src/javascripts/projects/common/modules/navigation/sticky.js (hunk 3 above)
Comment (OliverJAsh) [inline static/src/javascripts/projects/common/modules/navigation/sticky.js]: Not anymore https://github.com/guardian/frontend/pull/8493

**Pull request merged**